### DISCUSSION TITLE
Zephyr IP Stack Leaks in Promiscuous Mode

### DISCUSSION THREAD

**Discussion by bob2oneil** [zephyrproject-rtos/zephyr#45758] [General] [answered]
**Describe the bug**
I have a networking based application that provides 2 interfaces under the Native-POSIX board target, a simulated wired and simulated wireless interface.

The wired interface provides support for 2 UDP receive sockets, which can optionally (conditional compilation) be replaced with a singular raw layer 2 socket in a single thread.  The raw socket is the method that I need as I need Layer 2 support to be able to forward on ARPs and ICMPs across a network.

There is also a separate thread bound to a different UDP port.

A Telnet server is active on this interface to provide status and control.

In my current Zephyr configuration, when data is ingress into the raw socket (or the UDP sockets) in combination with the CONFIG_NET_PROMISCUOUS_MODE kernel configuration - the Zephyr stack appears to be leaking, resulting in an application failure.

This is confirmed by viewing the results of the shell output to "_net allocs_" command showing network buffer allocation and corresponding free (where present, lack of free shows the buffer leak during the clone procedure).

So it would seem that the stack allocations are not being properly de-registered and freed.

I am looking for some suggestions on where to dig into determine the problem and solution.

I can delay the amount of ingress traffic before the exception/failure by increasing the number of network buffers, but eventually it will happen.

I can get the leak to **not** occur if I disable **CONFIG_NET_PROMISCUOUS_MODE** in the kernel configuration, but then I can not pass Layer 2 traffic.

The problem is I need support for this and Layer 2 support in general using a raw socket to handle L2 traffic to
the primary wired interface to the application.  I need this support as well as a standard UDP socket and Telnet, all within the same application.  This setting makes a copy of the packet that is not being properly freed, but I am not
sure what mechanism would do the free.  I am reading the raw socket using the Linux/Zephyr standard 
multiplexed POSIX i/o using the **zephyr_select()** statement, wherein I would expect the Zephyr cloned packet
to be freed after the read occurs.

I am using the standard POSIX multiplexed i/o "select" call to handle the Ethernet interfaces similar to the following:

```
        // Wait for data to read or write between Ethernet interfaces or a timeout
        // NOTE: Need to set CONFIG_NET_SOCKETS_POLL_MAX to the total number of sockets to poll in select()
        int count = zsock_select(maxFD + 1, &readfds, 0, 0, (struct zsock_timeval *)&tvTimeout);
```

The read from the raw socket uses code as follows:
```

            if (m_wiredInterface.isOpenRawSocket())
            {
                sock = m_wiredInterface.getRawSocket();

                // Host wired interface reception of a packet
                if (ZSOCK_FD_ISSET(sock, &readfds))
                {
                    onWiredRawSocketReceive(tvNow, true);
                }
            }
```

The receive handler in POSIX is the following:

```
    // Receive packet from wired interface
    int bytesRead = receivePacket(m_wiredInterface.getDeviceName(),
                                  m_wiredInterface.getRawSocket(),
                                  m_wiredInterface.m_recvPacket,
                                  sizeof(m_wiredInterface.m_recvPacket),
                                  m_wiredInterface.m_recvPktSize);
```

Where:

```
int App::receivePacket(const char *ifName, const int sock, unsigned char *pkt, int pktSize, int &size)
{
    size = 0;

    // Reads packets from socket from wired or wirelesss interface
    int bytesRead = zsock_recv(sock, (void *)pkt, pktSize, 0);

    // Negative value indicates read error
    if (bytesRead <= 0)
    {
        LOG_ERR("zsock_recv(): error '%s' on interface %s, socket %d, bytes read %d", log_strdup(strerror(errno)), ifName, sock, bytesRead);
        return 0;
    }
    // Success on read
    else
    {
        LOG_DBG("zsock_recv() read %d bytes on interface %s on socket %d", bytesRead, log_strdup(ifName), sock);
    }

    // Assign count of bytes read
    size = bytesRead;

    // Return count of bytes read from the interface
    return bytesRead;
}
```

My expectation would be that after the POSIX API call to **zsock_recv**, any cloning of the network packet or dynamic memory would be automatically freed.  Using the POSIX APIs, I have no access to the underlying Zephyr specific buffers to allow the application to free the cloned packet.

Clearly I do not fully understand what is required by the application using this combination of a raw socket in promiscuous mode via the kernel configuration, a separate UPD socket in another thread, and a Telnet server to co-exist properly, and the de-referencing process that occurs to mark a network packet for deallocation.

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using?  Native-POSIX
 - What have you tried to diagnose or workaround this issue? Yes
 - ...

**To Reproduce**
1. west build -p always -b native_posix_64 ${workspaceFolder} 

**Expected behavior**
Network allocations are properly freed after the POSIX read when using a promiscuous raw socket.

**Impact**
Showstopper

**Logs and console output**

Here is a capture of the callstack at the point of the exception occurence.

```
net_pkt_cursor_init(struct net_pkt * pkt) (/home/robocom/zephyrproject/zephyr/subsys/net/ip/net_pkt.c:1516)
net_if_recv_data(struct net_if * iface, struct net_pkt * pkt) (/home/robocom/zephyrproject/zephyr/subsys/net/ip/net_if.c:3929)
process_data(_Bool is_loopback, struct net_pkt * pkt) (/home/robocom/zephyrproject/zephyr/include/zephyr/net/net_pkt.h:286)
processing_data(struct net_pkt * pkt, _Bool is_loopback) (/home/robocom/zephyrproject/zephyr/subsys/net/ip/net_core.c:141)
net_rx(struct net_if * iface, struct net_pkt * pkt) (/home/robocom/zephyrproject/zephyr/subsys/net/ip/net_core.c:365)
net_process_rx_packet(struct net_pkt * pkt) (/home/robocom/zephyrproject/zephyr/include/zephyr/net/net_pkt.h:286)
tc_rx_handler(struct k_fifo * fifo) (/home/robocom/zephyrproject/zephyr/subsys/net/ip/net_tc.c:250)
z_thread_entry(k_thread_entry_t entry, void * p1, void * p2, void * p3) (/home/robocom/zephyrproject/zephyr/lib/os/thread_entry.c:36)
posix_thread_starter(void * arg) (/home/robocom/zephyrproject/zephyr/arch/posix/core/posix_core.c:305)
libpthread.so.0!start_thread(void * arg) (/build/glibc-sMfBJT/glibc-2.31/nptl/pthread_create.c:477)
libc.so.6!clone() (/build/glibc-sMfBJT/glibc-2.31/sysdeps/unix/sysv/linux/x86_64/clone.S:95)
```

Here is a dump of "net allocs" from the Telnet interface menu with a simple 1 pulse per second PING traffic ingress to the primary Native-POSIX Ethernet interface.  Note that there is no corresponding free to the packet clone.

```
0x4f06d0/1       used      RX   net_pkt_clone_internal():1823
0x4f0658/1       used      RX   net_pkt_clone_internal():1823
0x4f05e0/1       used      RX   net_pkt_clone_internal():1823
0x4f0568/1       used      RX   prepare_non_vlan_pkt():305
0x4f04f0/1       used      RX   net_pkt_clone_internal():1823
0x4f0478/1       used      RX   net_pkt_clone_internal():1823
0x4f0400/1       used      RX   net_pkt_clone_internal():1823
0x4f0388/1       used      RX   net_pkt_clone_internal():1823
0x4f0310/1       used      RX   net_pkt_clone_internal():1823
0x4f0298/1       used      RX   net_pkt_clone_internal():1823
0x4f0220/1       used      RX   net_pkt_clone_internal():1823
0x4f01a8/1       used      RX   net_pkt_clone_internal():1823
0x4f0130/1       used      RX   net_pkt_clone_internal():1823
0x4f00b8/1       used      RX   net_pkt_clone_internal():1823
0x4f0040/1       used      RX   net_pkt_clone_internal():1823
0x4eff50/1       used      RX   net_pkt_clone_internal():1823
0x4efed8/1       used      RX   prepare_non_vlan_pkt():305
0x4effc8/1       used      RX   net_pkt_clone_internal():1823
0x4efde8/1       used      RX   net_pkt_clone_internal():1823
0x4efd70/1       used      RX   prepare_non_vlan_pkt():305
0x4efe60/1       used      RX   net_pkt_clone_internal():1823
0x4efc80/1       used      RX   net_pkt_clone_internal():1823
0x4efc08/1       used      RX   net_pkt_clone_internal():1823
0x4efb90/1       used      RX   net_pkt_clone_internal():1823
0x4efb18/1       used      RX   net_pkt_clone_internal():1823
0x4efaa0         free      RX   prepare_non_vlan_pkt():305 -> processing_data():159
0x4efcf8         free      RX   net_pkt_clone_internal():1823 -> zpacket_recvfrom_ctx():253

0x4f0040/1       used      RX   net_pkt_clone_internal():1823
0x4eff50/1       used      RX   net_pkt_clone_internal():1823
0x4efed8/1       used      RX   prepare_non_vlan_pkt():305
0x4effc8/1       used      RX   net_pkt_clone_internal():1823
0x4efde8/1       used      RX   net_pkt_clone_internal():1823
0x4efd70/1       used      RX   prepare_non_vlan_pkt():305
0x4efe60/1       used      RX   net_pkt_clone_internal():1823
0x4efaa0/1       used      RX   net_pkt_clone_internal():1823
0x4efc80/1       used      RX   net_pkt_clone_internal():1823
0x4efc08/1       used      RX   net_pkt_clone_internal():1823
0x4efb90/1       used      RX   net_pkt_clone_internal():1823
0x4efb18/1       used      RX   net_pkt_clone_internal():1823
0x4efcf8/1       used      RX   net_pkt_clone_internal():1823
0x4ef9b0/1       used      RX   net_pkt_clone_internal():1823
0x4ef938/1       used      RX   net_pkt_clone_internal():1823
0x4ef8c0/1       used      RX   prepare_non_vlan_pkt():305
0x4efa28/1       used      RX   net_pkt_clone_internal():1823
0x4dff60         free      TX   arp_prepare_reply():502 -> ethernet_send():717
0x4a9710         free   TDATA   ethernet_fill_header():504 -> ethernet_remove_l2_header():586
```

**Environment (please complete the following information):**
 - OS: (Linux)
 - Toolchain (e.g Zephyr SDK zephyr-sdk-0.13.1, .Booting Zephyr OS build v2.7.99-7456-g8566b00df3e1.)

**Additional context**

Project Kernel Configuration is as follows:

```
# Zephyr Project Kernel Configuration (common for all build board targets)

# C++ support
CONFIG_CPLUSPLUS=y
CONFIG_LIB_CPLUSPLUS=y
CONFIG_STD_CPP2A=y
CONFIG_CBPRINTF_FP_SUPPORT=y

# General kernel options
CONFIG_INIT_STACKS=y

# fs_dirent structures are big
CONFIG_MAIN_STACK_SIZE=2048

# Size of heap memory pool
CONFIG_HEAP_MEM_POOL_SIZE=16384

# Co-operative threading disabled, want pre-emption
# A cooperative thread remains the current thread until
# it performs an action that makes it unready
CONFIG_NET_TC_THREAD_COOPERATIVE=n

# Base 64 conversions support for XML interface
CONFIG_BASE64=y

# Flash storage settings
# Optionally force the file system to be recreated
CONFIG_FLASH=y
CONFIG_FLASH_MAP=y
CONFIG_FLASH_PAGE_LAYOUT=y
CONFIG_FILE_SYSTEM=y
CONFIG_FILE_SYSTEM_LITTLEFS=y
CONFIG_FS_LOG_LEVEL_DBG=y
# CONFIG_APP_WIPE_STORAGE=y

# Networking config
# Need TCP for Telnet session for shell interface
CONFIG_NETWORKING=y
CONFIG_NET_UDP=y
CONFIG_NET_TCP=y
CONFIG_NET_IPV6=n
CONFIG_NET_IPV4=y
CONFIG_NET_DHCPV4=n
CONFIG_NET_ARP=y
CONFIG_NET_NATIVE=y
CONFIG_NET_L2_ETHERNET=y
CONFIG_NET_DISABLE_ICMP_DESTINATION_UNREACHABLE=y

# Network drivers buffers
# Maximum is unsigned short as 65,535
CONFIG_NET_PKT_RX_COUNT=650
CONFIG_NET_PKT_TX_COUNT=650
CONFIG_NET_BUF_RX_COUNT=650
CONFIG_NET_BUF_TX_COUNT=650

# IP Address ranges
CONFIG_NET_IF_UNICAST_IPV4_ADDR_COUNT=258
CONFIG_NET_IF_MCAST_IPV4_ADDR_COUNT=0
CONFIG_NET_MAX_CONTEXTS=10

# To support prj.conf based IP setup
# We will be assigning IPs dynamically runtime
# so will not use the configuration based approach
CONFIG_NET_CONFIG_SETTINGS=n
CONFIG_NET_CONFIG_NEED_IPV6=n
CONFIG_NET_CONFIG_NEED_IPV4=n

# IP address options, leave blank so can be assigned runtime
# Primary interface IP and peer IP
CONFIG_NET_CONFIG_MY_IPV4_ADDR=""
CONFIG_NET_CONFIG_PEER_IPV4_ADDR=""

# Network interface statistics logging
CONFIG_NET_STATISTICS=y
CONFIG_NET_STATISTICS_USER_API=n
CONFIG_NET_STATISTICS_PERIODIC_OUTPUT=n
CONFIG_NET_STATISTICS_IPV4=y
CONFIG_NET_STATISTICS_IPV6=n
CONFIG_NET_STATISTICS_IPV6_ND=n
CONFIG_NET_STATISTICS_ICMP=y
CONFIG_NET_STATISTICS_UDP=y
CONFIG_NET_STATISTICS_TCP=y
CONFIG_NET_STATISTICS_MLD=n
CONFIG_NET_STATISTICS_ETHERNET=y
CONFIG_NET_STATISTICS_ETHERNET_VENDOR=n
CONFIG_NET_STATISTICS_LOG_LEVEL_DBG=y
CONFIG_NET_STATISTICS_PER_INTERFACE=y

# Enable promiscuous mode support on sockets for L2 access
# Necessary for Layer 2 Bridge behaviors (ARP and ICMP) on wired interface
CONFIG_NET_PROMISCUOUS_MODE=y

# Enable BSD sockets but don't use POSIX names for clashes with build machine
CONFIG_NET_SOCKETS=y
CONFIG_NET_SOCKETS_POSIX_NAMES=n

# Maximum file descriptors
CONFIG_POSIX_MAX_FDS=6

# Packet raw socket configuration to allow for raw sockets bound to iface
CONFIG_NET_SOCKETS_PACKET=y

# Random number generator
CONFIG_ENTROPY_GENERATOR=y

# Logging
CONFIG_NET_LOG=y
CONFIG_LOG=y
CONFIG_LOG_MODE_DEFERRED=n
CONFIG_LOG_RUNTIME_FILTERING=y
CONFIG_LOG_BUFFER_SIZE=2048
CONFIG_PRINTK=y
#CONFIG_LOG_MODE_OVERFLOW=y
#CONFIG_LOG_PROCESS_THREAD=y
#CONFIG_LOG_PROCESS_TRIGGER_THRESHOLD=10
#CONFIG_LOG_PROCESS_THREAD_STARTUP_DELAY_MS=0
#CONFIG_LOG_PROCESS_THREAD_SLEEP_MS=1000
#CONFIG_LOG_PROCESS_THREAD_STACK_SIZE=4096
#CONFIG_LOG_STRDUP_BUF_COUNT=128
#CONFIG_LOG_STRDUP_MAX_STRING=128
CONFIG_LOG_STRDUP_POOL_PROFILING=y
CONFIG_LOG_PROCESS_TRIGGER_THRESHOLD=0
CONFIG_LOG_STRDUP_BUF_COUNT=64

# - 0 OFF, do not write by default
# - 1 ERROR, default to only write LOG_LEVEL_ERR
# - 2 WARNING, default to write LOG_LEVEL_WRN
# - 3 INFO, default to write LOG_LEVEL_INFO
# - 4 DEBUG, default to write LOG_LEVEL_DBG
CONFIG_LOG_DEFAULT_LEVEL=3

# Prefix log output with function name
CONFIG_LOG_FUNC_NAME_PREFIX_ERR=y
CONFIG_LOG_FUNC_NAME_PREFIX_WRN=y
CONFIG_LOG_FUNC_NAME_PREFIX_INF=y
CONFIG_LOG_FUNC_NAME_PREFIX_DBG=y

# Disable logging using '%s' where strdup()s occur
# Only applies to logging using separate thread
CONFIG_LOG_DETECT_MISSED_STRDUP=n

# Additions to the Shell/Console Interface
# Note Telnet control of UART shell requires
# specifying the UART backend in the Telnet
# command (e.g. "log backend log_backend_uart status")
CONFIG_SHELL_STACK_SIZE=4096
CONFIG_SHELL_CMDS=y
CONFIG_SHELL_TAB_AUTOCOMPLETION=y
CONFIG_SHELL_HISTORY=y
CONFIG_NET_SHELL=y
CONFIG_KERNEL_SHELL=y
CONFIG_NET_STATISTICS=y
CONFIG_DEVICE_SHELL=y

# Shell and logging device backends
CONFIG_SHELL_LOG_BACKEND=n
CONFIG_SHELL_BACKEND_TELNET=y
CONFIG_LOG_BACKEND_UART=y
CONFIG_SHELL_TELNET_INIT_LOG_LEVEL_NONE=y

# Debug Options for debugging
CONFIG_DEBUG=y
CONFIG_DEBUG_OPTIMIZATIONS=y
#CONFIG_LOG_MODE_DEFERRED=n

# Memory buffer leak verification shell extensions
CONFIG_NET_BUF_POOL_USAGE=y
CONFIG_NET_DEBUG_NET_PKT_ALLOC=y

# File System Shell
CONFIG_FILE_SYSTEM=y
CONFIG_FILE_SYSTEM_SHELL=y

# Gather network statistics
CONFIG_NET_STATISTICS_ETHERNET=y

# Name threads
CONFIG_THREAD_NAME=y

# Max number of IPv4 network interfaces in the system
# Will be a value of 1 for real radio
CONFIG_NET_IF_MAX_IPV4_COUNT=2

# How many Tx traffic classes to have for each network device
# 1 = all the network traffic is handled equally.
# 0 = all the network traffic is pushed to the driver directly without any queues
CONFIG_NET_TC_TX_COUNT=1

# How many network connections is supported (default is 4)
CONFIG_NET_MAX_CONN=8

# Total number of sockets to poll in select()
CONFIG_NET_SOCKETS_POLL_MAX=5
```

**Comment by jukkar** [comment]
Are you familiar with the promiscuous mode sample app at `samples/net/promiscuous_mode`, it shows how to use the promiscuous mode API. In short, if you enable promiscuous mode by calling ` net_promisc_mode_on()`, then you need to have a handler that will need to unref the received packet, otherwise there is a leak.

**Comment by bob2oneil** [comment]
I understand, but this then implies that using standard POSIX APIs at the network layer is incompatible with promiscuous mode for the network interface, that is Zephyr is not compatible with that standard.  Is there any way that the underlying collection pointer to be freed could be exposed to say the zsock_recv() API so that the application could free it?, or would it just make sense that Zephyr perform the free itself as a result of that call?, meaning as soon as the content is copied to the user space buffer, there is no reason to keep the allocation around anymore.  I think this approach would be the most elegant, keeping Zephyr compatible to the POSIX standard and freeing the developer adhering to POSIX standard APIs out of the business of knowing deallocations.  Using the POSIX APIs provides efficiencies to folks porting existing applications, and adding promiscuous support would make Zephyr compatible with the Linux APIs.

**Comment by rlubos** [comment]
I think we've discussed this already in https://github.com/zephyrproject-rtos/zephyr/issues/36748.

TLDR the conclusion was that the module enabled with `CONFIG_NET_PROMISCUOUS_MODE` is a standalone API, not integrated with sockets, therefore I think it's safe to say that Zephyr does not support promiscuous RAW sockets. Threfore I think the issue should rather be cosidered as an enhancement, not a bug.

**Comment by bob2oneil** [comment]
Yes, the discussion from the past is the same one, over time I have found that I really needed that raw socket, which is not uncommon for any Layer 2 networking device, and as the Zephyr code base as it has evolved since my original posting, changes/updates have not addressed that issue.  

Is the level of effort described to have the **zsock_recv** API simply free up the local allocation a complex issue to implement?  

I really don't know the scope of what I am asking, and you guys being the subject matter experts, know the full scope of the design

**Comment by jukkar** [comment]
You could have a somewhat hackish workaround by having a "dummy" promiscuous mode handler that just discards the received packets, and then your raw socket should also receive all the same packets. This is not optimal of course but should work until we get support for proper promiscuous mode sockets.

**Comment by bob2oneil** [comment]
A workaround is just fine, I am not a purist.  How would this other handler synchronize packet removal and freeing relative to the raw socket implementation based on the zsock_select() multiplexed i/o and POSIX APIs?  Meaning, would this simply be a separate thread implementing the logic similar to the _promiscuous_mode_ sample similar to the following, where it waits on a packet, then frees it?  Would it be freeing a copy that would be used in the application thread implementing the POSIX communications APIs, i.e. would there need to be some form of thread synchronization?

```
	while (1) {
		pkt = net_promisc_mode_wait_data(K_FOREVER);
		net_pkt_unref(pkt);
	}
```

**Comment by jukkar** [comment, marked as answer]
Yeah, I was thinking a thread that would set the interface to promiscuous mode and have a while loop like you have above.
After some code digging there seems to be a simpler way. The promiscuous packets are processed in https://github.com/zephyrproject-rtos/zephyr/blob/bb8575962da291e3b75dfb3abe2fcb9341f76be2/subsys/net/ip/net_if.c#L3900 and if the interface is set to promiscuous mode but there is no handler i.e., the `net_promisc_mode_on()` is **not** called, then the packet should be unreffed properly and no leaks should happen.
So if you want to receive packets in promiscuous mode, but do not want to use the promiscuous mode API, you can enable promiscuous mode in config file, and then call `net_if_set_promisc(iface);` to set the interface to promiscuous mode. You should then be able to receive all the network traffic in your device without any buffer leaks and without any dummy handlers etc. Not tested but should work ok.

**Reply by bob2oneil** [reply]
Thanks Jukka, I will give your solution a try.    I want to make sure I understand your suggestion upon review.  If am already calling net_promisc_mode_on() in my code if the kernel configuration CONFIG_NET_PROMISCUOUS_MODE is enabled similar to the following:

```
    // Set interface to promiscuous mode to receive all packets on the wired interface
    // to support bridge operation with OTA ARPs and ICMPs
    if (IS_ENABLED(CONFIG_NET_PROMISCUOUS_MODE))
    {
        printf("Setting promiscuous mode on wired interface\n");

        // Promiscuous historically caused leaks with POSIX sockets, so may not be able to use it
        ifaceCallbackSetIfacePromisc(iface, 0);
    }

void App::ifaceCallbackSetIfacePromisc(struct net_if *iface, void *user_data)
{
    int ret = net_promisc_mode_on(iface);

    if (ret < 0)
    {
        if (ret == -EALREADY)
        {
            LOG_INF("Promiscuous mode already enabled");
            printk("Promiscuous mode already enabled\n");
        }
        else
        {
            LOG_ERR("Cannot set promiscuous mode for interface %p (%d)", iface, ret);
            printk("Cannot set promiscuous mode for interface %p (%d)\n", iface, ret);
        }
    }

    LOG_INF("Promiscuous mode set for interface %p (%d)", iface, ret);
    printk("Promiscuous mode set for interface %p (%d)\n", iface, ret);
}

```

I think what you are advocating for, is that I do NOT make this call, but instead call **net_if_set_promisc(iface)** while keeping CONFIG_NET_PROMISCUOUS_MODE in the kernel configuration file, so that the "clone" logic which follows per the following is NOT invoked.

```
enum net_verdict net_if_recv_data(struct net_if *iface, struct net_pkt *pkt)
{
	if (IS_ENABLED(CONFIG_NET_PROMISCUOUS_MODE) &&
	    net_if_is_promisc(iface)) {

```

It looks like the more primitive API **net_if_set_promisc(**) does not increment the enabled flag as follows:

```
int net_promisc_mode_on(struct net_if *iface)
{
	int ret;

	if (!iface) {
		return -EINVAL;
	}

	ret = net_if_set_promisc(iface);
	if (!ret) {
		atomic_inc(&enabled);
	}

	return ret;
}
```

I will need to test to make sure that this still allows the socket to receive raw Layer 2 packets such as ARPs and ICMPs.

The importance of promiscuous raw sockets is to implement a Layer 2 communications device, where I need to pass non-IPv4 traffic such as ARPs and ICMPs to implement a "bridge" type of device.  I can not think of a reason to have a raw Layer 2 socket without being promiscuous, and the two characteristics are often paired for bridge implementation purpose.

**Reply by bob2oneil** [reply]
I have tested out the Layer 2 functionality with the raw promiscuous socket, and preliminary testing is encouraging and I do not see the leak.  I want to pound on it a bit more ... but it surely looks fixed at the moment. 

If this were a bug report, I would close it.

Thanks Jukka and the entire Zephyr development team.

**Reply by IsikcanYilmaz** [reply]
Hello I'm a bit late to the game but i think im facing the exact same issue; im building a bridge type device with a w5500 shield and am simply just trying to get a hold of whatever bytes get sent my way so i can send them to the other side. no other logic. 

but i see on `net mem` that my buffers inflate and fill up eventually, despite me `unref`ing every packet that I receive. 

@bob2oneil could I ask how you solved this one? I used to use `net_promisc_mode_on(iface);` and i swapped it with `net_if_set_promisc(iface);` but now i get nothing thru the `net_promisc_mode_wait_data(K_FOREVER);` call and my buffer keeps filling up without freeing. 

also @jukkar if you could share hints/tips/tricks i would much appreciate.

many thanks!

**Comment by carlescufi** [comment]
> Yes, the discussion from the past is the same one, over time I have found that I really needed that raw socket, which is not uncommon for any Layer 2 networking device, and as the Zephyr code base as it has evolved since my original posting, changes/updates have not addressed that issue.
> 
> Is the level of effort described to have the **zsock_recv** API simply free up the local allocation a complex issue to implement?
> 
> I really don't know the scope of what I am asking, and you guys being the subject matter experts, know the full scope of the design

Given that you are really asking for a feature request here, I will convert this to a discussion and then ask you to open  a new issue (marked as enhancement), along the lines of: "Implement promiscuous mode for `AF_PACKET` sockets"
